File: ringstrings-app/src/test/java/com/coreman2200/ringstrings/symbol/inputprocessor/numerology/grouped/QualitiesProcessor.java
package com.coreman2200.ringstrings.symbol.inputprocessor.numerology.grouped;

import com.coreman2200.ringstrings.protos.RingStringsAppSettings;
import com.coreman2200.ringstrings.symbol.numbersymbol.grouped.BaseNumberSymbols;
import com.coreman2200.ringstrings.symbol.numbersymbol.grouped.GroupedNumberSymbols;
import com.coreman2200.ringstrings.symbol.numbersymbol.impl.GroupedNumberSymbolsImpl;
import com.coreman2200.ringstrings.symbol.numbersymbol.interfaces.IGroupedNumberSymbols;
import com.coreman2200.ringstrings.symbol.numbersymbol.interfaces.IListedNumberSymbols;
import com.coreman2200.ringstrings.symbol.numbersymbol.interfaces.INumberSymbol;
import com.coreman2200.ringstrings.symbol.numbersymbol.impl.ListedNumberSymbolsImpl;
import com.coreman2200.ringstrings.symbol.numbersymbol.grouped.Qualities;

import java.util.Arrays;
import java.util.HashMap;
import java.util.ArrayList;

/**
 * QualitiesProcessor
 * Processes the user's different numerological qualities.
 *
 * Created by Cory Higginbottom on 5/27/15
 * http://github.com/coreman2200
 *
 * Licensed under the GNU General Public License (GPL), Version 2.0.
 * You may not use this file except in compliance with this License.
 *
 * You may obtain a copy of the GPLv2 License at
 * http://www.gnu.org/licenses/old-licenses/gpl-2.0.html
 */

public class QualitiesProcessor extends GroupedNumberSymbolsInputProcessor implements IGroupedNumberSymbolsInputProcessor {
    IGroupedNumberSymbols processedQualities;

    public QualitiesProcessor(RingStringsAppSettings settings) {
        super(settings);
    }

    private String getFirstCharacterOfString(String text) {
        return text.substring(0, 1);
    }


    private INumberSymbol numGetBalance()
    {
        int firstNameFirstLetter = convertTextStringToValue(getFirstCharacterOfString(userProfile.getFirstName()));
        int midNameFirstLetter = convertTextStringToValue(getFirstCharacterOfString(userProfile.getMiddleName()));
        int lastNameFirstLetter = convertTextStringToValue(getFirstCharacterOfString(userProfile.getLastName()));

        return convertValueToNumberSymbol(firstNameFirstLetter + midNameFirstLetter + lastNameFirstLetter);
    }

    // TODO: ?? Not used..
    private INumberSymbol numGetBirthDay()
    {
        return convertValueToNumberSymbol(userProfile.getBirthDay());
    }

    private INumberSymbol numGetExpression()
    {
        int firstName = convertTextStringToValue(userProfile.getFirstName());
        int midName = convertTextStringToValue(userProfile.getMiddleName());
        int lastName = convertTextStringToValue(userProfile.getLastName());

        return convertValueToNumberSymbol(firstName + midName + lastName);
    }

    private INumberSymbol numGetLifePath()
    {
        int bday = userProfile.getBirthDay();
        int bmonth = userProfile.getBirthMonth();
        int byear = userProfile.getBirthYear();
        return convertValueToNumberSymbol(bday + bmonth + byear);
    }

    // Life Path + Expression..
    private INumberSymbol numGetMaturity()
    {
        int lifePath = processedQualities.getNumberSymbol(Qualities.LIFEPATH).getNumberSymbolValue();
        int expression = processedQualities.getNumberSymbol(Qualities.EXPRESSION).getNumberSymbolValue();
        assert (lifePath != 0);
        assert (expression != 0);

        return convertValueToNumberSymbol(lifePath + expression);
    }

    // First, Middle, Last..
    private INumberSymbol numGetPersonality()
    {
        int first = convertTextStringToValue(getConsonants(userProfile.getFirstName()));
        int middle = convertTextStringToValue(getConsonants(userProfile.getMiddleName()));
        int last = convertTextStringToValue(getConsonants(userProfile.getLastName()));

        return convertValueToNumberSymbol(first + middle + last);
    }

    // First Name, Bday..
    private INumberSymbol numGetRationalThought()
    {
        int i = convertTextStringToValue(userProfile.getFirstName()) + singularizeValue(userProfile.getBirthDay());
        return convertValueToNumberSymbol(i);
    }

    // First, Middle, Last name..
    private INumberSymbol numGetSoulUrge()
    {
        int firstName = convertTextStringToValue(getVowels(userProfile.getFirstName()));
        int midName = convertTextStringToValue(getVowels(userProfile.getMiddleName()));
        int lastName = convertTextStringToValue(getVowels(userProfile.getLastName()));

        return convertValueToNumberSymbol(firstName + midName + lastName);
    }

    // Karmic Lessons Count..
    private INumberSymbol numGetSubconsciousSelf()
    {
        IGroupedNumberSymbols karmiclessons = processedQualities.getGroupedNumberSymbol(Qualities.KARMICLESSON);

        int karmiccount = karmiclessons.size();
        return convertValueToNumberSymbol(9 - karmiccount);
    }

    private INumberSymbol numGetHiddenPassionsFromOccurrenceMap(HashMap<BaseNumberSymbols, Integer> occurrences) {
        final ArrayList<INumberSymbol> passions = getListOfMostCommonFromMap(occurrences);

        if (passions.size() == 0)
            return BaseNumberSymbols.ZERO.getBaseNumberSymbol();

        IListedNumberSymbols hpgroup = new ListedNumberSymbolsImpl(GroupedNumberSymbols.HIDDENPASSIONS);

        for (INumberSymbol lesson : passions)
            hpgroup.addNumberSymbol(lesson);

        return hpgroup;
    }


    private HashMap<BaseNumberSymbols, Integer> numMapOccurenceOfNumberSymbolsInName() {
        final String fullname = userProfile.getFirstName() + userProfile.getMiddleName() + userProfile.getLastName();
        char[] nameArray = prepareTextForProcessing(fullname);
        HashMap<BaseNumberSymbols, Integer> occurrences = new HashMap<>();

        for (char c : nameArray) {
            int charValue = mNumberSystem.numValueForChar(c);
            BaseNumberSymbols symbol = BaseNumberSymbols.getBaseNumberSymbolIDForValue(charValue);
            Integer count = (!occurrences.containsKey(symbol)) ? Integer.valueOf(0) : occurrences.get(symbol);
            occurrences.put(symbol, ++count);
        }
        return occurrences;
    }

    private ArrayList<INumberSymbol> getListOfMostCommonFromMap(HashMap<BaseNumberSymbols, Integer> occurrences) {

        Integer[] counts = occurrences.values().toArray(new Integer[occurrences.size()]);
        Arrays.sort(counts);

        int high = counts[counts.length-1];

        ArrayList<INumberSymbol> mostcommon = new ArrayList<>();

        for (BaseNumberSymbols symbol : BaseNumberSymbols.values()) {
            if (occurrences.containsKey(symbol) && occurrences.get(symbol) == high)
                mostcommon.add(symbol.getBaseNumberSymbol());
        }

        return mostcommon;
    }

    private INumberSymbol numGetKarmicLessonsListFromOccurrenceMap(HashMap<BaseNumberSymbols, Integer> occurrences) {
        final ArrayList<INumberSymbol> lessons = getListOfNotOccurredFromMap(occurrences);

        IListedNumberSymbols klgroup = new ListedNumberSymbolsImpl(GroupedNumberSymbols.KARMICLESSON);

        for (INumberSymbol lesson : lessons)
            klgroup.addNumberSymbol(lesson);

        return klgroup;
    }

    private ArrayList<INumberSymbol> getListOfNotOccurredFromMap(HashMap<BaseNumberSymbols, Integer> occurrences) {
        ArrayList<INumberSymbol> notfound = new ArrayList<>();
        for (BaseNumberSymbols symbol : BaseNumberSymbols.values()) {
            if (symbol.equals(BaseNumberSymbols.ZERO))
                continue;
            if (!occurrences.containsKey(symbol) && !checkIsSymbolMasterNumber(symbol))
                notfound.add(symbol.getBaseNumberSymbol());
        }
        return notfound;
    }

    protected final boolean checkIsSymbolMasterNumber(BaseNumberSymbols symbol) {
        assert (symbol != null);
        return symbol.compareTo(BaseNumberSymbols.NINE) > 0;
    }

    @Override
    public IGroupedNumberSymbols produceGroupedNumberSymbolsForProfile() {
        // TODO: Protocol for building grouped & charted elems..
        processedQualities = new GroupedNumberSymbolsImpl(GroupedNumberSymbols.QUALITIES);
        processedQualities.addNumberSymbol(Qualities.LIFEPATH, numGetLifePath());
        processedQualities.addNumberSymbol(Qualities.BALANCE, numGetBalance());
        processedQualities.addNumberSymbol(Qualities.EXPRESSION, numGetExpression());
        processedQualities.addNumberSymbol(Qualities.MATURITY, numGetMaturity());
        processedQualities.addNumberSymbol(Qualities.RATIONALTHOUGHT, numGetRationalThought());
        processedQualities.addNumberSymbol(Qualities.SOULURGE, numGetSoulUrge());
        processedQualities.addNumberSymbol(Qualities.PERSONALITY, numGetPersonality());

        HashMap<BaseNumberSymbols, Integer> occurrences = numMapOccurenceOfNumberSymbolsInName();
        processedQualities.addNumberSymbol(Qualities.HIDDENPASSIONS, numGetHiddenPassionsFromOccurrenceMap(occurrences));
        processedQualities.addNumberSymbol(Qualities.KARMICLESSON, numGetKarmicLessonsListFromOccurrenceMap(occurrences));

        processedQualities.addNumberSymbol(Qualities.SUBCONSCIOUSSELF, numGetSubconsciousSelf());

        return processedQualities;
    }
}

File: domain/src/main/java/com/coreman2200/ringstrings/domain/input/numerology/grouped/QualitiesProcessor.kt
package com.coreman2200.ringstrings.domain.input.numerology.grouped
import com.coreman2200.ringstrings.domain.NumerologySettings
import com.coreman2200.ringstrings.domain.swisseph.IProfileData
import com.coreman2200.ringstrings.domain.symbol.SymbolModel
import com.coreman2200.ringstrings.domain.symbol.numbersymbol.grouped.BaseNumbers
import com.coreman2200.ringstrings.domain.symbol.numbersymbol.grouped.GroupedNumbers
import com.coreman2200.ringstrings.domain.symbol.numbersymbol.grouped.Qualities
import com.coreman2200.ringstrings.domain.symbol.numbersymbol.impl.BaseNumberSymbol
import com.coreman2200.ringstrings.domain.symbol.numbersymbol.impl.GroupedNumberSymbol
import com.coreman2200.ringstrings.domain.symbol.numbersymbol.interfaces.INumberSymbol
import java.util.*

/**
 * QualitiesProcessor
 * Processes the user's different numerological qualities.
 *
 * Created by Cory Higginbottom on 5/27/15
 * http://github.com/coreman2200
 *
 * Licensed under the GNU General Public License (GPL), Version 2.0.
 * You may not use this file except in compliance with this License.
 *
 * You may obtain a copy of the GPLv2 License at
 * http://www.gnu.org/licenses/old-licenses/gpl-2.0.html
 */
class QualitiesProcessor(
    profile: IProfileData,
    settings: NumerologySettings
) : GroupedNumberSymbolsInputProcessor(
    profile, settings), IGroupedNumberSymbolsInputProcessor {
    private val processed: GroupedNumberSymbol = GroupedNumberSymbol(GroupedNumbers.QUALITIES)
    private fun getFirstCharacterOfString(text: String): String {
        return text.substring(0, 1)
    }

    private fun numGetBalance(): INumberSymbol {
        val firstNameFirstLetter =
            convertTextStringToValue(getFirstCharacterOfString(profile.firstName()))
        val midNameFirstLetter =
            convertTextStringToValue(getFirstCharacterOfString(profile.middleName()))
        val lastNameFirstLetter =
            convertTextStringToValue(getFirstCharacterOfString(profile.lastName()))
        return convertValueToNumberSymbol(firstNameFirstLetter + midNameFirstLetter + lastNameFirstLetter)
    }

    // TODO: ?? Not used..
    private fun numGetBirthDay(): INumberSymbol {
        return convertValueToNumberSymbol(profile.getBirthDay())
    }

    private fun numGetExpression(): INumberSymbol {
        val firstName = convertTextStringToValue(profile.firstName())
        val midName = convertTextStringToValue(profile.middleName())
        val lastName = convertTextStringToValue(profile.lastName())
        return convertValueToNumberSymbol(firstName + midName + lastName)
    }

    private fun numGetLifePath(): INumberSymbol {
        val bday = profile.getBirthDay()
        val bmonth = profile.getBirthMonth()
        val byear = profile.getBirthYear()
        return convertValueToNumberSymbol(bday + bmonth + byear)
    }

    // Life Path + Expression..
    private fun numGetMaturity(): INumberSymbol {
        val lifePath = processed.get(Qualities.LIFEPATH) as INumberSymbol
        val expression = processed.get(Qualities.EXPRESSION) as INumberSymbol
        return convertValueToNumberSymbol(lifePath.value + expression.value)
    }

    // First, Middle, Last..
    private fun numGetPersonality(): INumberSymbol {
        val first = convertTextStringToValue(getConsonants(profile.firstName()))
        val middle = convertTextStringToValue(getConsonants(profile.middleName()))
        val last = convertTextStringToValue(getConsonants(profile.lastName()))
        return convertValueToNumberSymbol(first + middle + last)
    }

    // First Name, Bday..
    private fun numGetRationalThought(): INumberSymbol {
        val i =
            convertTextStringToValue(profile.firstName()) + singularizeValue(profile.getBirthDay())
        return convertValueToNumberSymbol(i)
    }

    // First, Middle, Last name..
    private fun numGetSoulUrge(): INumberSymbol {
        val firstName = convertTextStringToValue(getVowels(profile.firstName()))
        val midName = convertTextStringToValue(getVowels(profile.middleName()))
        val lastName = convertTextStringToValue(getVowels(profile.lastName()))
        return convertValueToNumberSymbol(firstName + midName + lastName)
    }

    // Karmic Lessons Count..
    private fun numGetSubconsciousSelf(): INumberSymbol {
        val karmiclessons: GroupedNumberSymbol = processed.get(Qualities.KARMICLESSON) as GroupedNumberSymbol
        val karmiccount = karmiclessons.size
        return convertValueToNumberSymbol(9 - karmiccount)
    }

    private fun numGetHiddenPassionsFromOccurrenceMap(occurrences: Map<BaseNumbers, Int>): SymbolModel {
        val passions = getListOfMostCommonFromMap(occurrences)
        if (passions.size == 0) return BaseNumberSymbol(BaseNumbers.ZERO)
        val hpgroup = GroupedNumberSymbol(GroupedNumbers.HIDDENPASSIONS)
        for (lesson in passions) hpgroup.add(lesson)
        return hpgroup
    }

    private fun numMapOccurenceOfNumberSymbolsInName(): Map<BaseNumbers, Int> {
        val fullname = profile.firstName() + profile.middleName() + profile.lastName()
        val nameArray = prepareTextForProcessing(fullname)
        val occurrences = mutableMapOf<BaseNumbers, Int>()
        for (c in nameArray) {
            val charValue = numSystem.numValueForChar(c)
            val symbol = BaseNumbers.values()[charValue]
            occurrences[symbol]?.plus(1)
        }
        return occurrences
    }

    private fun getListOfMostCommonFromMap(occurrences: Map<BaseNumbers, Int>): ArrayList<SymbolModel> {
        val counts = occurrences.values.toTypedArray()
        Arrays.sort(counts)
        val high = counts[counts.size - 1]
        val mostcommon = ArrayList<SymbolModel>()
        for (symbol in BaseNumbers.values()) {
            if (occurrences.containsKey(symbol) && occurrences[symbol] == high) mostcommon.add(
                BaseNumberSymbol(symbol)
            )
        }
        return mostcommon
    }

    private fun numGetKarmicLessonsListFromOccurrenceMap(occurrences: Map<BaseNumbers, Int>): GroupedNumberSymbol {
        val lessons = getListOfNotOccurredFromMap(occurrences)
        val klgroup = GroupedNumberSymbol(GroupedNumbers.KARMICLESSON)
        for (lesson in lessons) klgroup.add(lesson)
        return klgroup
    }

    private fun getListOfNotOccurredFromMap(occurrences: Map<BaseNumbers, Int>): ArrayList<BaseNumberSymbol> {
        val notfound = ArrayList<BaseNumberSymbol>()
        for (symbol in BaseNumbers.values()) {
            if (symbol == BaseNumbers.ZERO) continue
            if (!occurrences.containsKey(symbol) && !checkIsSymbolMasterNumber(symbol)) notfound.add(
                BaseNumberSymbol(symbol)
            )
        }
        return notfound
    }

    protected fun checkIsSymbolMasterNumber(symbol: BaseNumbers): Boolean {
        return symbol.value() > 9
    }

    override fun produceGroupedNumberSymbolsForProfile(): GroupedNumberSymbol {
        processed.clear()
        processed.add(Qualities.LIFEPATH, numGetLifePath() as SymbolModel)
        processed.add(Qualities.BALANCE, numGetBalance() as SymbolModel)
        processed.add(Qualities.EXPRESSION, numGetExpression() as SymbolModel)
        processed.add(Qualities.MATURITY, numGetMaturity() as SymbolModel)
        processed.add(Qualities.RATIONALTHOUGHT, numGetRationalThought() as SymbolModel)
        processed.add(Qualities.SOULURGE, numGetSoulUrge() as SymbolModel)
        processed.add(Qualities.PERSONALITY, numGetPersonality() as SymbolModel)
        val occurrences = numMapOccurenceOfNumberSymbolsInName()
        processed.add(
            Qualities.HIDDENPASSIONS,
            numGetHiddenPassionsFromOccurrenceMap(occurrences)
        )
        processed.add(
            Qualities.KARMICLESSON,
            numGetKarmicLessonsListFromOccurrenceMap(occurrences)
        )
        processed.add(Qualities.SUBCONSCIOUSSELF, numGetSubconsciousSelf() as SymbolModel)
        return processed
    }
}
